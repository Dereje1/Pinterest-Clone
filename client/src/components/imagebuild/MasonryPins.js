import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Masonry from 'react-masonry-component';
import HandleThumbnailImage from './HandleThumbnailImage';

const MasonryPins = ({
  layoutComplete,
  pinEnlarge,
  onBrokenImage,
  pinImage,
  deletePin,
  pins,
}) => {
  const [loadedImages, setLoadedImages] = useState([]);

  return (
    <Masonry
      onImagesLoaded={layoutComplete}
      className="my-gallery-class"
      options={{ fitWidth: true }}
    >
      {
        pins.map(element => (
          <div
            key={element._id}
            role="button"
            className="image-box"
            onClick={e => pinEnlarge(e, element)}
            onKeyDown={() => {}}
            tabIndex={0}
          >
            <img
              alt={element.imgDescription}
              onError={() => onBrokenImage(element._id)}
              className="image-format"
              src={element.imgLink}
              onLoad={() => setLoadedImages([...loadedImages, element._id])}
              style={{ visibility: loadedImages.includes(element._id) ? 'visible' : 'hidden' }}
            />
            <div className="description">
              {element.imgDescription}
            </div>
            <HandleThumbnailImage
              element={element}
              pinImage={pinImage}
              deletePin={deletePin}
            />
            <div className="owner">{`${element.owner.name}`}</div>
          </div>
        ))
      }
    </Masonry>
  );
};
export default MasonryPins;

MasonryPins.defaultProps = {
  pins: [null],
  pinImage: null,
  deletePin: null,
};

MasonryPins.propTypes = {
  pinImage: PropTypes.func,
  deletePin: PropTypes.func,
  layoutComplete: PropTypes.func.isRequired,
  pinEnlarge: PropTypes.func.isRequired,
  onBrokenImage: PropTypes.func.isRequired,
  pins: PropTypes.arrayOf(PropTypes.any),
};
