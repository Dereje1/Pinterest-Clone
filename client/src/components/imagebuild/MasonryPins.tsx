import React, { useState } from 'react';
import Masonry from 'react-masonry-component';
import { Typography } from '@mui/material';
import HandleThumbnailImage from './HandleThumbnailImage';
import { PinType } from '../../interfaces';

interface MasonryPinsProps {
  pins: PinType[]
  layoutComplete: () => void
  pinEnlarge: (e: React.SyntheticEvent, pin: PinType) => void
  onBrokenImage: (_id: string) => void
  pinImage: (pin: PinType) => void
  deletePin: ((pin: PinType) => void) | null
}

const initialLoadedImages: string[] = [];

function MasonryPins({
  layoutComplete,
  pinEnlarge,
  onBrokenImage,
  pinImage,
  deletePin,
  pins,
}: MasonryPinsProps) {
  const [loadedImages, setLoadedImages] = useState(initialLoadedImages);

  return (
    <Masonry
      onImagesLoaded={layoutComplete}
      className="my-gallery-class"
      options={{ fitWidth: true }}
    >
      {
        pins.map((pin) => (
          <div
            key={pin._id}
            role="button"
            className="image-box"
            onClick={(e) => pinEnlarge(e, pin)}
            onKeyDown={() => ({})}
            tabIndex={0}
          >
            <img
              alt={pin.imgDescription}
              onError={() => onBrokenImage(pin._id)}
              className="image-format"
              src={pin.imgLink}
              onLoad={() => setLoadedImages([...loadedImages, pin._id])}
              style={{ visibility: loadedImages.includes(pin._id) ? 'visible' : 'hidden' }}
            />
            <div className="description">
              <Typography variant="h6" sx={{ fontWeight: 'bold' }}>{`${pin.imgDescription}`}</Typography>
              <Typography variant="h6" sx={{ fontWeight: 'bold' }}>{`ðŸ“Œ ${pin.savedBy.length}`}</Typography>
            </div>
            <HandleThumbnailImage
              pin={pin}
              pinImage={pinImage}
              deletePin={deletePin}
            />
            <div className="owner">{`${pin.owner.name}`}</div>
          </div>
        ))
      }
    </Masonry>
  );
}
export default MasonryPins;
