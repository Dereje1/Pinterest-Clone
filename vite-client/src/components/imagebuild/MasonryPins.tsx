import React, { useState } from 'react';
import Masonry from 'react-masonry-component';
import { Typography } from '@mui/material';
import HandleThumbnailImage from './HandleThumbnailImage';
import { PinType } from '../../interfaces';

interface MasonryPinsProps {
  pins: PinType[]
  layoutComplete: () => void
  pinEnlarge: (e: React.SyntheticEvent, pin: PinType) => void
  onBrokenImage: (_id: string) => void
  pinImage: (pin: PinType) => void
  deletePin: ((pin: PinType) => void) | null
}

const initialLoadedImages: string[] = [];

function MasonryPins({
  layoutComplete,
  pinEnlarge,
  onBrokenImage,
  pinImage,
  deletePin,
  pins,
}: MasonryPinsProps) {
  const [loadedImages, setLoadedImages] = useState(initialLoadedImages);

  return (
    <Masonry
      onImagesLoaded={layoutComplete}
      className="my-gallery-class"
      options={{ fitWidth: true }}
    >
      {
        pins.map((element) => (
          <div
            key={element._id}
            role="button"
            className="image-box"
            onClick={(e) => pinEnlarge(e, element)}
            onKeyDown={() => ({})}
            tabIndex={0}
          >
            <img
              alt={element.imgDescription}
              onError={() => onBrokenImage(element._id)}
              className="image-format"
              src={element.imgLink}
              onLoad={() => setLoadedImages([...loadedImages, element._id])}
              style={{ visibility: loadedImages.includes(element._id) ? 'visible' : 'hidden' }}
            />
            <div className="description">
              <Typography variant="h6" sx={{ fontWeight: 'bold' }}>{`${element.imgDescription}`}</Typography>
              <Typography variant="h6" sx={{ fontWeight: 'bold' }}>{`ðŸ“Œ ${element.savedBy.length}`}</Typography>
            </div>
            <HandleThumbnailImage
              element={element}
              pinImage={pinImage}
              deletePin={deletePin}
            />
            <div className="owner">{`${element.owner.name}`}</div>
          </div>
        ))
      }
    </Masonry>
  );
}
export default MasonryPins;
