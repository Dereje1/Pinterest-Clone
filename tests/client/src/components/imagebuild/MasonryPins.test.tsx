/**
 * @jest-environment jsdom
 */
import React from 'react';
import { EnzymePropSelector, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import MasonryPins from '../../../../../client/src/components/imagebuild/MasonryPins';
import { pinsStub } from '../../../stub';

describe('The MasonryPins component', () => {
  let props;
  beforeEach(() => {
    props = {
      pinImage: jest.fn(),
      deletePin: jest.fn(),
      layoutComplete: jest.fn(),
      pinEnlarge: jest.fn(),
      onBrokenImage: jest.fn(),
      pins: pinsStub,
    };
  });

  afterEach(() => {
    props = null;
  });

  test('will render....', () => {
    const wrapper = shallow(<MasonryPins {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will send parameters to enlarge a pin', () => {
    const wrapper = shallow(<MasonryPins {...props} />);
    const imageBox = wrapper.find({ className: 'image-box' }).at(0);
    imageBox.props().onKeyDown();
    imageBox.props().onClick('event');
    expect(props.pinEnlarge).toHaveBeenCalledWith('event', pinsStub[0]);
  });

  test('will set the visibility of a pin if it is done loading', () => {
    const wrapper = shallow(<MasonryPins {...props} />);
    let imageBox = wrapper.find({ className: 'image-box' }).at(0);
    let img: EnzymePropSelector = imageBox.find('img');
    expect(img.props().style.visibility).toBe('hidden');
    img.props().onLoad();
    imageBox = wrapper.find({ className: 'image-box' }).at(0);
    img = imageBox.find('img');
    expect(img.props().style.visibility).toBe('visible');
  });

  test('will send parameters for an error in loading', () => {
    const wrapper = shallow(<MasonryPins {...props} />);
    const imageBox: EnzymePropSelector = wrapper.find({ className: 'image-box' }).at(0);
    const img = imageBox.find('img');
    img.props().onError();
    expect(props.onBrokenImage).toHaveBeenCalledWith(1);
  });
});
