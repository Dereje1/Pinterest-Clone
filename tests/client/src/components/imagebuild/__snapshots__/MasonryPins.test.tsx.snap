// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The MasonryPins component will render.... 1`] = `
<MasonryComponent
  className="my-gallery-class"
  disableImagesLoaded={false}
  elementType="div"
  enableResizableChildren={false}
  imagesLoadedOptions={{}}
  onImagesLoaded={[MockFunction]}
  onLayoutComplete={[Function]}
  onRemoveComplete={[Function]}
  options={
    {
      "fitWidth": true,
    }
  }
  updateOnEachImageLoad={false}
>
  <div
    className="image-box"
    key="1"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <img
      alt="imgDescription id-1"
      className="image-format"
      onError={[Function]}
      onLoad={[Function]}
      src="imgLink id-1"
      style={
        {
          "visibility": "hidden",
        }
      }
    />
    <div
      className="description"
    >
      <ForwardRef(Typography)
        sx={
          {
            "fontWeight": "bold",
          }
        }
        variant="h6"
      >
        imgDescription id-1
      </ForwardRef(Typography)>
      <ForwardRef(Typography)
        sx={
          {
            "fontWeight": "bold",
          }
        }
        variant="h6"
      >
        ðŸ“Œ 2
      </ForwardRef(Typography)>
    </div>
    <HandleThumbnailImage
      deletePin={[MockFunction]}
      pin={
        {
          "_id": "1",
          "comments": [
            {
              "_id": "Comment_id_1",
              "comment": "Test comment",
              "createdAt": "2021-07-28T16:30:48.572Z",
              "displayName": "tester-displayName",
              "userId": "Comment_User_Id_1",
            },
          ],
          "createdAt": "2020-12-25T00:00:00.000+00:00",
          "hasSaved": false,
          "imgDescription": "imgDescription id-1",
          "imgLink": "imgLink id-1",
          "owner": {
            "name": "owner id-1",
            "service": "google",
            "userId": "1",
          },
          "owns": false,
          "savedBy": [
            {
              "name": "savedBy - id-1",
              "service": "google",
              "userId": "1",
            },
            {
              "name": "savedBy - id-1",
              "service": "github",
              "userId": "1",
            },
          ],
          "tags": [
            {
              "_id": "tag_id_1",
              "tag": "tag 1",
            },
            {
              "_id": "tag_id_2",
              "tag": "tag 2",
            },
          ],
        }
      }
      pinImage={[MockFunction]}
    />
    <div
      className="owner"
    >
      owner id-1
    </div>
  </div>
  <div
    className="image-box"
    key="2"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <img
      alt="imgDescription id-2"
      className="image-format"
      onError={[Function]}
      onLoad={[Function]}
      src="imgLink id-2"
      style={
        {
          "visibility": "hidden",
        }
      }
    />
    <div
      className="description"
    >
      <ForwardRef(Typography)
        sx={
          {
            "fontWeight": "bold",
          }
        }
        variant="h6"
      >
        imgDescription id-2
      </ForwardRef(Typography)>
      <ForwardRef(Typography)
        sx={
          {
            "fontWeight": "bold",
          }
        }
        variant="h6"
      >
        ðŸ“Œ 2
      </ForwardRef(Typography)>
    </div>
    <HandleThumbnailImage
      deletePin={[MockFunction]}
      pin={
        {
          "_id": "2",
          "comments": [],
          "createdAt": "2020-12-25T00:00:00.000+00:00",
          "hasSaved": true,
          "imgDescription": "imgDescription id-2",
          "imgLink": "imgLink id-2",
          "owner": {
            "name": "owner id-2",
            "service": "twitter",
            "userId": "2",
          },
          "owns": false,
          "savedBy": [
            {
              "name": "savedBy - id-2",
              "service": "google",
              "userId": "2",
            },
            {
              "name": "savedBy - id-2git",
              "service": "github",
              "userId": "2",
            },
          ],
          "tags": [
            {
              "_id": "tag_id_1",
              "tag": "tag 1",
            },
          ],
        }
      }
      pinImage={[MockFunction]}
    />
    <div
      className="owner"
    >
      owner id-2
    </div>
  </div>
  <div
    className="image-box"
    key="3"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <img
      alt="imgDescription id-3"
      className="image-format"
      onError={[Function]}
      onLoad={[Function]}
      src="imgLink id-3"
      style={
        {
          "visibility": "hidden",
        }
      }
    />
    <div
      className="description"
    >
      <ForwardRef(Typography)
        sx={
          {
            "fontWeight": "bold",
          }
        }
        variant="h6"
      >
        imgDescription id-3
      </ForwardRef(Typography)>
      <ForwardRef(Typography)
        sx={
          {
            "fontWeight": "bold",
          }
        }
        variant="h6"
      >
        ðŸ“Œ 2
      </ForwardRef(Typography)>
    </div>
    <HandleThumbnailImage
      deletePin={[MockFunction]}
      pin={
        {
          "_id": "3",
          "comments": [],
          "createdAt": "2020-12-25T00:00:00.000+00:00",
          "hasSaved": false,
          "imgDescription": "imgDescription id-3",
          "imgLink": "imgLink id-3",
          "owner": {
            "name": "owner id-3",
            "service": "github",
            "userId": "3",
          },
          "owns": true,
          "savedBy": [
            {
              "name": "savedBy - id-3",
              "service": "google",
              "userId": "3",
            },
            {
              "name": "savedBy - id-3",
              "service": "github",
              "userId": "3",
            },
          ],
          "tags": [],
        }
      }
      pinImage={[MockFunction]}
    />
    <div
      className="owner"
    >
      owner id-3
    </div>
  </div>
</MasonryComponent>
`;
